### PR TITLE
ENH: add quantile method to resample

### LINKED ISSUE #15023: ENH: add quantile method to resample
#### Problem description

The `quantile` method is currently not supported on `resample`. While `resample(...).agg(lambda x: x.quantile(..))` works fine, it would be nice to add the `resample(..).quantile(..)` shortcut.

#### Code Sample

```
In [19]: ts = pd.Series(range(20), index=pd.date_range('2016-01-01', periods=20))

In [21]: ts.resample('W').agg(lambda x: x.quantile(0.75))
Out[21]: 
2016-01-03     1.5
2016-01-10     7.5
2016-01-17    14.5
2016-01-24    18.5
Freq: W-SUN, dtype: float64
```

So it would be nice the following would give the same result:
```
In [22]: ts.resample('W').quantile(0.75)
/home/joris/miniconda3/envs/dev/bin/ipython:1: FutureWarning: 
.resample() is now a deferred operation
You called quantile(...) on this deferred object which materialized it into a series
by implicitly taking the mean.  Use .resample(...).mean() instead
  #!/home/joris/miniconda3/envs/dev/bin/python
Out[22]: 14.25
```

The fact that this currently implicitly takes the mean before calculating the quantile (`ts.resample('W').mean().quantile(0.75)`) would make this change slightly API breaking.

Using pandas master, 0.19.0+289.g1bf94c8

### PR DESCRIPTION
**Pull request by discort**
- [x] closes #15023
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.24.0.txt [hunk 1]
```diff
@@ -650,6 +650,7 @@ Groupby/Resample/Rolling
   ``SeriesGroupBy`` when the grouping variable only contains NaNs and numpy version < 1.13 (:issue:`21956`).
 - Multiple bugs in :func:`pandas.core.Rolling.min` with ``closed='left'` and a
   datetime-like index leading to incorrect results and also segfault. (:issue:`21704`)
+- Added :meth:`Resampler.quantile` (:issue:`15023`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: move to other enhancements

File: pandas/core/resample.py [hunk 2]
```diff
@@ -749,6 +749,16 @@ def size(self):
             result = pd.Series([], index=result.index, dtype='int64')
         return result
 
+    def quantile(self, q=0.5, **kwargs):
+        """
+        Return value at the given quantile.
+
```
Comment (jreback) [inline pandas/core/resample.py]: add a versionadded tag

File: doc/source/whatsnew/v0.24.0.txt (hunk 1 above)
Comment (jreback) [reply doc/source/whatsnew/v0.24.0.txt]: this needs to be added in api.rst as well

File: pandas/tests/test_resample.py [hunk 3]
```diff
@@ -41,7 +41,7 @@
 
 # The various methods we support
 downsample_methods = ['min', 'max', 'first', 'last', 'sum', 'mean', 'sem',
-                      'median', 'prod', 'var', 'ohlc']
```
Comment (jreback) [inline pandas/tests/test_resample.py]: can you add the test from the OP which compares 2 methods of getting the same result

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by discort** [comment]
@jreback

File: pandas/core/resample.py [hunk 4]
```diff
@@ -749,6 +749,18 @@ def size(self):
...
+        Parameters
+        ----------
+        q : float or array-like, default 0.5 (50% quantile)
+
```
Comment (jreback) [inline pandas/core/resample.py]: can you add some See Also and point to the Series / groupby versions

File: pandas/tests/test_resample.py [hunk 5]
```diff
@@ -2165,6 +2165,13 @@ def test_resample_datetime_values(self):
         res = df['timestamp'].resample('2D').first()
         tm.assert_series_equal(res, exp)
 
+    def test_resample_quantile(self):
```
Comment (jreback) [inline pandas/tests/test_resample.py]: can you test on timedelta / period as well. you might want to move to the Base class to make this easier.

File: doc/source/whatsnew/v0.24.0.txt [hunk 6]
```diff
@@ -182,6 +182,7 @@ Other Enhancements
 - :func:`to_timedelta` now supports iso-formated timedelta strings (:issue:`21877`)
 - :class:`Series` and :class:`DataFrame` now support :class:`Iterable` in constructor (:issue:`2193`)
 - :class:`DatetimeIndex` gained :attr:`DatetimeIndex.timetz` attribute. Returns local time with timezone information. (:issue:`21358`)
+- Added :meth:`Resampler.quantile` (:issue:`15023`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: can you say that  ``Series.resample`` and ``DataFrame.resample`` have gained the ``Resample.quantile()`` method (use references for .resample)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/test_resample.py (hunk 5 above)
Comment (discort) [reply pandas/tests/test_resample.py]: This example mentioned in the description of issue doesn't work for timedelta:

```
s = self.create_series()
expected = s.resample('W').agg(lambda x: x.quantile(0.75))
```

trace:
```
.../pandas/pandas/tseries/offsets.py in nanos(self)
    399     @property
    400     def nanos(self):
--> 401         raise ValueError("{name} is a non-fixed frequency".format(name=self))
    402
    403

ValueError: <Week: weekday=6> is a non-fixed frequency
```

@jreback 
Let me know please if it's correct.

File: pandas/core/resample.py [hunk 7]
```diff
@@ -1053,9 +1059,15 @@ def _downsample(self, how, **kwargs):
         how = self._is_cython_func(how) or how
         ax = self.ax
 
+        # Empty groupby objs are accepted only callable funcs
```
Comment (jreback) [inline pandas/core/resample.py]: what triggers this?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/resample.py (hunk 7 above)
Comment (jreback) [reply pandas/core/resample.py]: hmm this is odd. can you remove this fix, and instead: create a new issue which demonstrates this (is this *only* with this method?) and xfail the test. I think something bigger is going on here. We can merge this then come back with a battery of tests for this.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/resample.py (hunk 7 above)
Comment (discort) [reply pandas/core/resample.py]: @jreback 

to more precisely, this [block](https://github.com/pandas-dev/pandas/blob/ac64c8e6049b438e8f59c129115c450f226c5ffa/pandas/core/groupby/groupby.py#L668-L677) doesn't pass the check and raises exception.
It became to fail after adding `kwargs` to [_groupby_and_aggregate](https://github.com/pandas-dev/pandas/pull/22304/files#diff-28a4430519696cd7934b5e6a5c54feb6R1070)

Comment (jreback) [reply pandas/core/resample.py]: hmm, this is old logic. can you try calling 

``def _try_aggregate_string_function(self, arg, *args, **kwargs)`` (its defined in the base class) and see if this works?

Comment (discort) [reply pandas/core/resample.py]: No, it doesn't

```
.../pandas/pandas/core/groupby/generic.py in _cython_agg_general(self, how, alt, numeric_only, min_count)
     74                             min_count=-1):
     75         new_items, new_blocks = self._cython_agg_blocks(
---> 76             how, alt=alt, numeric_only=numeric_only, min_count=min_count)
     77         return self._wrap_agged_blocks(new_items, new_blocks)
     78

.../pandas/pandas/core/groupby/generic.py in _cython_agg_blocks(self, how, alt, numeric_only, min_count)
    147
    148         if len(new_blocks) == 0:
--> 149             raise DataError('No numeric types to aggregate')
    150
    151         # reset the locs in the blocks to correspond to our

DataError: No numeric types to aggregate
```

Comment (jreback) [reply pandas/core/resample.py]: hmm, ok, then let's just xfail this test and pls open another issue. I think this is a bigger issues and needs to be addressed separately.

Comment (discort) [reply pandas/core/resample.py]: @jreback 

Here is the issue - https://github.com/pandas-dev/pandas/issues/22339

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/test_resample.py [hunk 8]
```diff
@@ -771,6 +771,15 @@ def test_apply_to_empty_series(self):
 
             assert_series_equal(result, expected, check_dtype=False)
```
[the remark this answers is not in the document]
Comment (discort) [reply pandas/tests/test_resample.py]: The second way looks simpler. So I closed new PR/issue

**Comment by discort** [comment]
@jreback

File: pandas/core/resample.py [hunk 9]
```diff
@@ -749,6 +749,24 @@ def size(self):
...
+        ----------
+        q : float or array-like, default 0.5 (50% quantile)
+
+        See Also
+        --------
+        Series.quantile
+        DataFrame.quantile
+        DataFrameGroupBy.quantile
+
+        .. versionadded:: 0.24.0
```
Comment (jreback) [inline pandas/core/resample.py]: the tag goes right below the summary line (the first line of the doc-string, with a blank line before and after)

**Review by jreback — CHANGES_REQUESTED** [review]
minor comment on the doc-string, ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @discort keep em coming!

**Pull request merged**